## Import Library

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path
from sentence_transformers import SentenceTransformer
import torch
from annoy import AnnoyIndex
%matplotlib inline
%reload_ext autoreload
%autoreload 2

## Import Data

In [2]:
dataPath = Path("../data/shopify/")
apps_df = pd.read_csv(dataPath/"shopify_apps.csv")
apps_df.head(10)

app_id  \
0  9e4748a9-7eda-4814-83b6-0537d44152b1   
1  d1476138-a608-4bb9-8d39-b30f3ca7617d   
2  d6e49a3c-2f9f-4bfa-8c26-5d024faf2241   
3  0ef0087f-3ae5-4dbc-84e0-193b576d82ed   
4  7aac2a1f-ff03-4f38-aeb7-7619403a6f05   
5  c13bfb7f-8b5a-40c6-a338-dbdec5cfd130   
6  6a71634f-f94f-498d-8713-d2b01ca90917   
7  da50f0bf-d116-46a0-b0c1-0d90c05d8ffe   
8  89735c3c-4d25-40f2-8b54-150f59cfb099   
9  3c89c108-4858-4b07-893a-460ea5a0d91a   

                                                 url  \
0                 https://apps.shopify.com/translate   
1        https://apps.shopify.com/instant-brand-page   
2                  https://apps.shopify.com/powr-faq   
3            https://apps.shopify.com/mps-promote-me   
4  https://apps.shopify.com/installify-boost-mobi...   
5         https://apps.shopify.com/easy-product-feed   
6           https://apps.shopify.com/posbill-connect   
7              https://apps.shopify.com/bulk-fulfill   
8         https://apps.shopify.com/show-price-in-btc   
9              https://apps.shopify.com/product-copy   

                           title  \
0       Panda Language Translate   
1             Instant Brand Page   
2    FAQ Accordion | Help Center   
3  Promote Me | Many apps in one   
4                      Instalify   
5              EASY product feed   
6                PosBill Connect   
7                   Bulk Fulfill   
8              Show Price in BTC   
9        POKY ‑ Product Importer   

                                         description  \
0  Panda Language Translation improve your busine...   
1  Having a brand page gives your shoppers a quic...   
2  Create, organize and display Frequently Asked ...   
3  Promote Me app is a bundle of apps that includ...   
4  We are trusted by hundreds of Shopify and Shop...   
5  There are a lot of product feed apps out there...   
6  PosBill Connect - the perfection connection to...   
7  Upload to fulfill: \n Bulk Fulfill by Upatra a...   
8  Simply convert your store's prices into ANY cu...   
9  Poky \n Allows you to import/copy products fro...   

                                             tagline  \
0       Translate your store into multiple languages   
1         A-Z Brand Index Page and Favourites Slider   
2  FAQ page, FAQ accordion menu for product Info ...   
3  Spin Wheel,Currency Converter,Quick ATC Button...   
4               Supercharge Your Mobile App Installs   
5  Get your products listed on Google Shopping th...   
6  Connect your Webshop to the solutions of PosBill.   
7  Reduce wasted time - Automatically sync your t...   
8  A better currency converter app. Premium featu...   
9  Copy / Import products from any Shop store wit...   

                                         category  
0                                    Store design  
1                                    Store design  
2                   Store design,Customer support  
3  Store design,Sales and conversion optimization  
4                                    Store design  
5                                       Marketing  
6        Orders and shipping,Inventory management  
7                Orders and shipping,Productivity  
8                                    Store design  
9        Productivity,Finding and adding products

## Step1: Encoding the App description

We are going to use app descriptions to find similar Shopify apps. In order to do that we need to create emeddings from the description using transformer model. We will be using sentence-transformers/paraphrase-xlm-r-multilingual-v1 model using sentence-transformers library.

In [5]:
## Initiate Model
model = SentenceTransformer('sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2')
model.to(torch.device("cuda"))

SentenceTransformer(
  (0): Transformer({'max_seq_length': 128, 'do_lower_case': False}) with Transformer model: BertModel 
  (1): Pooling({'word_embedding_dimension': 384, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False})
)

In [6]:
## Getting Embedding for app shopify apps
embeddings = model.encode(apps_df.description.astype(str).to_list())
print(embeddings.shape)

(4750, 384)


In [23]:
## Creating App2idx 
app2idx = {app_id:idx for idx,app_id in enumerate(apps_df.app_id)}

This model returns embedding of size 384

## Step 2: Define parameters and build Annoy Index

In [17]:
f = embeddings.shape[1] # Length of item vector that will be indexed
t = AnnoyIndex(f, 'angular')  ## Defining index
for idx, vector in enumerate(embeddings):
    t.add_item(idx, vector)

## Caculating trees
t.build(int(np.round(np.sqrt(len(embeddings))))) # Number of trees = sqrt(length of embedding vector)
t.save(str(dataPath/'appEmbedding.ann')) ## saving the annoy index

True

## Step 3:  Search similar apps

In [29]:
def get_similar_apps(app_id, app2idx, app_data, annoy_index, n_items=5):
    query_idx = app2idx[app_id]
    idxs = annoy_index.get_nns_by_item(query_idx, n_items+1)
    return app_data.iloc[idxs,:]

### Some Examples

In [30]:
## Langugage translation apps
get_similar_apps(app_id="9e4748a9-7eda-4814-83b6-0537d44152b1" , app2idx = app2idx, app_data = apps_df, annoy_index=t, n_items=5)

app_id  \
0     9e4748a9-7eda-4814-83b6-0537d44152b1   
893   be566fe3-2449-4ae5-8b59-06b79a668dff   
2409  6de8f41b-f0f6-4e3c-b5b4-731b833741e3   
1396  cccdbfd7-b1fd-4079-8b49-280e19f3a8b9   
1892  6281aca5-c6f7-4858-a445-813ec7214abc   
4268  9b6804b6-91d8-4813-a353-f5080155fe56   

                                                    url  \
0                    https://apps.shopify.com/translate   
893       https://apps.shopify.com/mokool-translate-pro   
2409   https://apps.shopify.com/easy-language-translate   
1396        https://apps.shopify.com/translate-my-store   
1892  https://apps.shopify.com/language-translate-by...   
4268                   https://apps.shopify.com/langify   

                            title  \
0        Panda Language Translate   
893    Translate Pro Using Google   
2409      Easy Language Translate   
1396           Translate My Store   
1892  Uplinkly Language Translate   
4268                      langify   

                                            description  \
0     Panda Language Translation improve your busine...   
893   Do you want both more customers and to increas...   
2409  Translate your store and attract global visito...   
1396  Limited Time Offer for the next 500 users: You...   
1892  Click ‘Add App’ to start using the translation...   
4268  How does langify improve the business of your ...   

                                                tagline      category  
0          Translate your store into multiple languages  Store design  
893   Auto Translate to 100 languages in fast. Unlim...  Store design  
2409  Experience a better way to translate store and...  Store design  
1396  Translate your store, increase sales in global...  Store design  
1892  Translate your store & get more sales automati...  Store design  
4268        Translate your shop into multiple languages  Store design

The top row is the query indexed and bottom 5 rows are the similar apps returned by annoy. Looks like translatation app is returning translations

In [36]:
query_id = apps_df.loc[np.random.randint(0, len(apps_df)),'app_id'] ##randomly choosing an app_id
get_similar_apps(app_id=query_id , app2idx = app2idx, app_data = apps_df, annoy_index=t, n_items=5) ## Query for similar apps

app_id  \
1896  d957681d-0ceb-4226-bde4-75a5ff4cc10d   
2498  f3b34865-523a-4f88-8e69-95023eed9a37   
3847  f718b743-2b12-40d5-8935-db20587ea678   
2479  1ee3ad7a-248e-4b32-bb0b-0b9ba76b888d   
4471  c4359733-0471-47ab-8959-bf6bd569181e   
3521  76f39b1f-1c05-4fb0-9616-dd508132349d   

                                                   url  \
1896                https://apps.shopify.com/evmupsell   
2498               https://apps.shopify.com/dnt-upsell   
3847           https://apps.shopify.com/in-cart-upsell   
2479  https://apps.shopify.com/upseller-product-upsell   
4471   https://apps.shopify.com/product-upsell-manager   
3521         https://apps.shopify.com/checkout-booster   

                               title  \
1896  Discounted Upsell Bundle, Bogo   
2498          Easy Upsell Cross‑sell   
3847                  In Cart Upsell   
2479        SellUp • Free Upsell App   
4471     Upsell & Cross‑Sell Booster   
3521       Upsell + Related Products   

                                            description  \
1896  ==== # Officially the Sellify Apps Alternative...   
2498  Upsell and Cross-sell product app on Shopify b...   
3847  Targeted, Popup-Free Upsells \n In Cart Upsell...   
2479  Increase your sales and up your upsell game wi...   
4471  Features \n \n This app gives you the ability ...   
3521  4 in 1 Sales Booster App (with Upsell cross-se...   

                                                tagline  \
1896  Upsellio - In cart upsell, buy x get y, BOGO &...   
2498  Increase sales with upsell, cross-sell and mak...   
3847  Laser targeted upsell offers in the shopping c...   
2479  Upsell, Bundles, Discounts, Cross-sell & One C...   
4471  Can show targeted products as an offer to rele...   
3521  Upsell Popup, Recently Viewed Products, BOGO O...   

                                            category  
1896               Sales and conversion optimization  
2498               Sales and conversion optimization  
3847  Store design,Sales and conversion optimization  
2479               Sales and conversion optimization  
4471               Sales and conversion optimization  
3521                                    Store design

Upsell Related apps are returning upsell

In [37]:
query_id = apps_df.loc[np.random.randint(0, len(apps_df)),'app_id'] ##randomly choosing an app_id
get_similar_apps(app_id=query_id , app2idx = app2idx, app_data = apps_df, annoy_index=t, n_items=5) ## Query for similar apps

app_id  \
1101  5557170e-c5db-44b4-82a5-31db79a7363d   
3806  5a37f0f7-0090-4750-b4d5-806dd1d74179   
939   ccd7adc7-2430-459e-96c2-25ba91cbaf87   
516   0fe33ede-d70c-4369-af6e-d5e6d7d39949   
1374  d305c7f0-7d20-40e3-be85-f97c809cd4c1   
4729  ac6fa8f3-bc49-40c9-9cbb-f974322c799a   

                                                    url  \
1101      https://apps.shopify.com/predictive-preloader   
3806             https://apps.shopify.com/plug-in-speed   
939            https://apps.shopify.com/avada-seo-suite   
516   https://apps.shopify.com/jumbo-page-speed-opti...   
1374          https://apps.shopify.com/page-speed-boost   
4729            https://apps.shopify.com/imageoptimizer   

                               title  \
1101  Instant Page Speeds AI Powered   
3806         Plug in Speed optimizer   
939   AVADA SEO Image Optimizer FREE   
516     Jumbo ‑ Page speed optimizer   
1374  Page Speed+SEO Image Optimizer   
4729                 Image Optimizer   

                                            description  \
1101  We use machine learning AI to predict where us...   
3806  Click "Add App" for a free trial of the code a...   
939   We guarantee to contribute this app for FREE F...   
516   Jumbo Shopify Page Speed Optimizer. \n Jumbo o...   
1374  High-quality images make your store a nice app...   
4729  Speed up Your Store, Improve SEO and User Expe...   

                                                tagline  \
1101  Increase Sales, Conv. Rates, Reduce Bounce, 1-...   
3806  Compress and Optimize Images, CSS & Javascript...   
939   FREE FOREVER GUARANTEE! SEO Image Optimizer, A...   
516   Page speed optimizer for improved conversion a...   
1374  Image Compression, Site Speed Optimization, Pa...   
4729  Image Optimizer: Image Compression and Image S...   

                    category  
1101            Store design  
3806               Marketing  
939   Store design,Marketing  
516             Store design  
1374            Store design  
4729            Store design

Page seats optimizer apps are returning right results